horovod/common/operations.cc

```diff
@@ -808,72 +810,108 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
 
   Status status;
   if (response.response_type() == MPIResponse::ALLGATHER) {
-    assert(entries.size() == 1);
-    auto e = entries[0];
-
-    // Copy tensor sizes from the MPI response into a vector of int64_t
-    // and compute total size.  This is size of first dimension.
     std::vector<int64_t> tensor_sizes;
-    int64_t total_dimension_size = 0;
-    for (auto sz : response.tensor_sizes()) {
-      tensor_sizes.push_back(sz);
-      total_dimension_size += sz;
-    }
 
-    // Every tensor participating in Allgather operation may have different
-    // first dimension size, but the rest of dimensions are same for all
-    // tensors.  Here we get shape of tensor sliced by first dimension.
-    TensorShape single_slice_shape;
-    for (int i = 1; i < e.tensor->shape().dims(); ++i) {
-      single_slice_shape.AddDim(e.tensor->shape().dim_size(i));
+    // Sizes of subcomponents of each entry from all ranks
+    auto** entry_component_sizes = new int64_t*[entries.size()];
+
+    // Offset of each subcomponent of every entry in the final buffer after
+    // allgatherv
+    auto** entry_component_offsets = new int64_t*[entries.size()];
+
+    auto* recvcounts = new int[horovod_global.size]();
+    auto* displcmnts = new int[horovod_global.size]();
+
+    for (int ec = 0; ec < entries.size(); ec++) {
+      entry_component_sizes[ec] = new int64_t[horovod_global.size]();
+      entry_component_offsets[ec] = new int64_t[horovod_global.size]();
     }
 
-    // Allgather output will have shape of:
-    // (sum of first dimension of every tensor) x (tensor slice shape).
-    TensorShape output_shape;
-    output_shape.AddDim((int64_t)total_dimension_size);
-    output_shape.AppendShape(single_slice_shape);
+    auto& first_entry = entries[0];
 
     ACTIVITY_START_ALL(entries, timeline, ALLOCATE_OUTPUT)
-    status = e.context->AllocateOutput(output_shape, &e.output);
-    if (!status.ok()) {
-      timeline.End(e.tensor_name, nullptr);
-      e.callback(status);
-      return;
+    for (unsigned int ec = 0; ec < entries.size(); ++ec) {
+      auto& e = entries[ec];
+      // Every tensor participating in Allgather operation may have different
+      // first dimension size, but the rest of dimensions are same for all
+      // tensors.  Here we get shape of tensor sliced by first dimension.
+      TensorShape single_slice_shape;
+      for (int i = 1; i < e.tensor->shape().dims(); ++i) {
+        single_slice_shape.AddDim(e.tensor->shape().dim_size(i));
+      }
+
+      // Copy tensor sizes from the MPI response into a vector of int64_t
+      // and compute total size.  This is size of first dimension.
+      int64_t total_entry_dimension_size = 0;
+      for (unsigned int rc = 0; rc < horovod_global.size; ++rc) {
+        tensor_sizes.push_back(
+            response.tensor_sizes()[ec * horovod_global.size + rc]);
+        total_entry_dimension_size +=
```

Comment:
Introduce local variable for `response.tensor_sizes()[ec * horovod_global.size + rc]` ?